README.md
EVENT TIMELINE
==============

The Event Timeline (http://vaadin.com/directory#addon/event-timeline) is a widget for the vaadin web-development framework. 
It is inspired by the NASA OSTPV (Onboard Short Term Plan Viewer) application to display mission plans and timelines.

The widget is derived from the Vaadin Timeline (http://vaadin.com/directory#addon/vaadin-timeline), and modified to support events rather than data graphs. 

Features
--------

+ Support for multiple independent event bands
+ Overlapping events in different stripes
+ Adjust band heights dynamically
+ Custom styling of event labels with CSS
+ Easy to use time-line browser with a draggable and zoomable interface
+ Customizable zoom levels

Usage
-----

To use the Event Timeline, just add a TimelineEventProvider as an event band: 


``` java
import java.util.Calendar;
import java.util.Date;

import com.spaceapplications.vaadin.addon.eventtimeline.EventTimeline;
import com.spaceapplications.vaadin.addon.eventtimeline.event.BasicEvent;
import com.spaceapplications.vaadin.addon.eventtimeline.event.BasicEventProvider;
import com.spaceapplications.vaadin.addon.eventtimeline.event.TimelineEventProvider;
import com.vaadin.Application;
import com.vaadin.ui.Window;

public class SimpleEventTimelineDemo extends Application {
  @Override
  public void init() {
    Window mainWindow = new Window("Simple EventTimeline Demo");

    setMainWindow(mainWindow);

    // create the timeline
    EventTimeline timeline = new EventTimeline("Our event timeline");
    timeline.setHeight("500px");
    timeline.setWidth("100%");

    // set the visible time range
    Calendar cal = Calendar.getInstance();
    Date start = cal.getTime();
    cal.add(Calendar.HOUR_OF_DAY, 24);
    Date end = cal.getTime();
    timeline.setVisibleDateRange(start, end);

    // add our data sources
    timeline.addEventBand("Band", createEventProvider(end));

    // Add some zoom levels
    timeline.addZoomLevel("Hour", 60 * 60 * 1000L);
    timeline.addZoomLevel("Day", 86400000L);

    mainWindow.addComponent(timeline);
  }

  public TimelineEventProvider createEventProvider(final Date end) {
    BasicEventProvider provider = new BasicEventProvider();

    // get events for a whole day
    Calendar cal = Calendar.getInstance();
    cal.add(Calendar.HOUR, 1);
    int idx = 0;
    while (cal.getTime().before(end)) {
      // create a simple event
      BasicEvent event = new BasicEvent();
      event.setEventId(String.valueOf(idx));
      
      // set the timestamp property
      event.setStart(cal.getTime());
      cal.add(Calendar.MINUTE, 10);
      event.setEnd(cal.getTime());
      // set the caption
      event.setCaption("Event");
      // style the event
      event.setStyleName("color1");

      provider.addEvent(event);
      cal.add(Calendar.MINUTE, 30);      
    }

    return provider;
  }  
}
```

Copyright and license
---------------------

Copyright 2009-2011 Vaadin Ltd., 2011 Space Applications Services NV/SA.

This program is available under GNU Affero General Public License 
(version 3 or later at your option).

See the file licensing.txt distributed with this software for more
information about licensing.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

CHANGELOG.txt
                                     EVENT TIMELINE
                                  Development Changelog

Version 0.6.0 - 11.04.2013
--------------------------

   - improved initialization procedure (reduced steps and requests)
   - fixed lock selection when data updates are received
   - added getEvents(String) method to TimelineEventProvider
   - remove band only once, contributed by Florian Pirchner
   
Version 0.5.1 - 17.01.2012
--------------------------

   - re-sending event bands after browser reload

Version 0.5.0 - 17.01.2012
--------------------------

   - added support for dynamically adding / removing bands (fixes issue #7)
       contributed by Florian Pirchner
   - added support for band selection (fixes issue #9)
       contributed by Florian Pirchner
   - added support for paging of event bands
       contributed by Florian Pirchner
   - event label description can now be any HTML content (fixes issue #2)
   - added support for explicitly setting the band height (fixes issue #8)
   
Version 0.4.0 - 08.12.2011
--------------------------

   - added maven build config
   - band adjuster is only visible when hovering over the band area
   - the grid color can now be set using CSS3 style rgba values
   - bands automatically expand to maximum available height
   - added first support for event updates
   - events are correctly sorted according their start and end times (fixes issue #5)
   - improved positioning of band adjuster (fixes issue #6)
 
Version 0.3.0 - 16.09.2011
--------------------------

   - initial release to vaadin addon-store and github


demo/com/spaceapplications/vaadin/addon/eventtimeline/demo/MyEventTimelineDemo.java
package com.spaceapplications.vaadin.addon.eventtimeline.demo;

import java.util.Calendar;
import java.util.Date;

import com.spaceapplications.vaadin.addon.eventtimeline.EventTimeline;
import com.spaceapplications.vaadin.addon.eventtimeline.EventTimeline.EventButtonClickEvent;
import com.spaceapplications.vaadin.addon.eventtimeline.event.BasicEvent;
import com.spaceapplications.vaadin.addon.eventtimeline.event.BasicEventProvider;
import com.spaceapplications.vaadin.addon.eventtimeline.event.TimelineEventProvider;
import com.vaadin.Application;
import com.vaadin.ui.Label;
import com.vaadin.ui.Window;

@SuppressWarnings("serial")
public class MyEventTimelineDemo extends Application {
  @Override
  public void init() {
    Window mainWindow = new Window("EventTimeline Demo Application");
    Label label = new Label("Hello Vaadin user");
    mainWindow.addComponent(label);
    setMainWindow(mainWindow);

    // Create the timeline
    EventTimeline timeline = new EventTimeline("Our event timeline");
    timeline.setHeight("500px");
    timeline.setWidth("100%");

    Calendar cal = Calendar.getInstance();
    Date start = cal.getTime();
    cal.add(Calendar.DAY_OF_WEEK, 1);
    Date end = cal.getTime();
    
    // Create the data sources
    final TimelineEventProvider eventProvider1 = createEventProvider1(end);
    final TimelineEventProvider eventProvider2 = createEventProvider2(end);
    final TimelineEventProvider eventProvider3 = createEventProvider3(end);

    // Add our data sources
    timeline.addEventBand("Band 1", eventProvider1);
    timeline.addEventBand("Band 2", eventProvider2);
    timeline.addEventBand("Band 3", eventProvider3);

    timeline.setVisibleDateRange(start, end);

    // Add some zoom levels
    timeline.addZoomLevel("Hour", 60 * 60 * 1000L);
    timeline.addZoomLevel("3Hour", 3 * 60 * 60 * 1000L);
    timeline.addZoomLevel("6Hour", 6 * 60 * 60 * 1000L);
    timeline.addZoomLevel("12Hour", 12 * 60 * 60 * 1000L);
    timeline.addZoomLevel("Day", 86400000L);
    //timeline.addZoomLevel("Week", 7 * 86400000L);
    // timeline.addZoomLevel("Month", 2629743830L);

    mainWindow.addComponent(timeline);
    
    // Listen to click events from events
    timeline.addListener(new EventTimeline.EventClickListener() {
      public void eventClick(EventButtonClickEvent event) {
        getMainWindow().showNotification("clicked event " + event.getItemId());
        
        BasicEvent e = (BasicEvent) eventProvider1.getEvent((String) event.getItemId());
        if (e != null) {
          e.setCaption("Test");
        }
      }
    });
  }

  public TimelineEventProvider createEventProvider1(final Date end) {
    BasicEventProvider provider = new BasicEventProvider();
    
    Calendar cal = Calendar.getInstance();
    cal.add(Calendar.HOUR, 6);
    int idx = 0;
    while (cal.getTime().before(end)) {
      // Create a point in time
      BasicEvent event = new BasicEvent();
      event.setEventId(String.valueOf(idx));
      
      // Set the timestamp property
      event.setStart(cal.getTime());
      cal.add(Calendar.MINUTE, 3);
      event.setEnd(cal.getTime());
      event.setCaption("Event");

      provider.addEvent(event);
      
      if (idx++ % 2 == 0) {
        cal.add(Calendar.MINUTE, 5);
        event.setStyleName("color1");
      } else {
        cal.add(Calendar.MINUTE, -2);
        event.setStyleName("color2");
      }
    }

    return provider;
  }
  
  public TimelineEventProvider createEventProvider2(final Date end) {
    BasicEventProvider provider = new BasicEventProvider();
    
    Calendar cal = Calendar.getInstance();
    cal.add(Calendar.HOUR_OF_DAY, 2);
    int idx = 0;
    while (cal.getTime().before(end)) {
      // Create a point in time
      BasicEvent event = new BasicEvent();
      event.setEventId(String.valueOf(idx++));
      
      // Set the timestamp property
      event.setStart(cal.getTime());
      cal.add(Calendar.MINUTE, 10);
      event.setEnd(cal.getTime());
      event.setCaption("Task");
      event.setStyleName("color4");

      provider.addEvent(event);
      cal.add(Calendar.HOUR, 1);
    }

    return provider;
  }
  
  public TimelineEventProvider createEventProvider3(final Date end) {
    BasicEventProvider provider = new BasicEventProvider();
    
    Calendar cal = Calendar.getInstance();
    cal.add(Calendar.HOUR, 6);
    int idx = 0;
    while (cal.getTime().before(end)) {
      // Create a point in time
      BasicEvent event = new BasicEvent();
      event.setEventId(String.valueOf(idx));
      
      // Set the timestamp property
      event.setStart(cal.getTime());
      cal.add(Calendar.MINUTE, 30);
      event.setEnd(cal.getTime());
      event.setCaption("Event");
      event.setDescription("Test");

      provider.addEvent(event);
      
      idx++;
      if (idx % 3 == 0) {
        cal.add(Calendar.MINUTE, 30);
        event.setStyleName("color3");
      } else if (idx % 3 == 1){
        cal.add(Calendar.MINUTE, -20);
        event.setStyleName("color4");
      } else {
        cal.add(Calendar.MINUTE, -8);
        event.setStyleName("color2");
      }
    }

    return provider;
  }  
}